driver-core/src/main/com/mongodb/internal/operation/MixedBulkWriteOperation.java

```diff
@@ -180,32 +180,22 @@ public BulkWriteResult execute(final WriteBinding binding) {
             logRetryExecute(retryState);
             return withSourceAndConnection(binding::getWriteConnectionSource, true, (source, connection) -> {
                 ConnectionDescription connectionDescription = connection.getDescription();
-                int maxWireVersion = connectionDescription.getMaxWireVersion();
                 // attach `maxWireVersion` ASAP because it is used to check whether we can retry
-                retryState.attach(AttachmentKeys.maxWireVersion(), maxWireVersion, true);
-                BulkWriteTracker bulkWriteTracker = retryState.attachment(AttachmentKeys.bulkWriteTracker())
-                        .orElseThrow(Assertions::fail);
+                retryState.attach(AttachmentKeys.maxWireVersion(), connectionDescription.getMaxWireVersion(), true);
                 SessionContext sessionContext = binding.getSessionContext();
                 WriteConcern writeConcern = getAppliedWriteConcern(sessionContext);
-                if (!retryState.isFirstAttempt() && !isRetryableWrite(retryWrites, writeConcern, source.getServerDescription(),
-                        connectionDescription, sessionContext)) {
-                    RuntimeException prospectiveFailedResult = (RuntimeException) retryState.exception().orElse(null);
-                    retryState.breakAndThrowIfRetryAnd(() -> !(prospectiveFailedResult instanceof MongoWriteConcernWithResponseException));
-                    bulkWriteTracker.batch().ifPresent(bulkWriteBatch -> {
-                        assertTrue(prospectiveFailedResult instanceof MongoWriteConcernWithResponseException);
-                        bulkWriteBatch.addResult((BsonDocument) ((MongoWriteConcernWithResponseException) prospectiveFailedResult)
-                                .getResponse());
-                        BulkWriteTracker.attachNext(retryState, bulkWriteBatch);
-                    });
```

Comment:
This `if` (except for the `!isRetryableWrite` condition) was extracted to `handleMongoWriteConcernWithResponseException`, and then the new method was used in one more place, such that now it's called both before and after an attempt, as opposed to be called only before. That's what the fix is about.

So now `handleMongoWriteConcernWithResponseException` is called both before and after an `executeBulkWriteBatch` attempt (this is also how it was before the retrying mechanism was refactored). Maybe we can refactor `MixedBulkWriteOperation` such that there is only one call to `handleMongoWriteConcernWithResponseException`, but I did not see how.